src/libstore/build/derivation-goal.cc

```diff
@@ -3121,6 +3121,13 @@ void DerivationGoal::registerOutputs()
             newInfo0.references = refs.second;
             if (refs.first)
                 newInfo0.references.insert(newInfo0.path);
+            if (scratchPath != newInfo0.path) {
+                // Also rewrite the output path
```

Comment:
Ah it just occurred to me, `refs.first` is whether we have a self reference, so we can use this to optimize slightly.

I also threw in a comment explaining. Feel free to reword.
```suggestion
            if (refs.first) {
                // patch has self-references, so should also rewrite them.
                // Note that this doesn't change the end-goal CA we ca calculated above, since that calculation skips the contents of self-references.
                // But rather, if we have self-references, relocate from `scratchPath`, and then *don't* do this,
                // our self-references to `scratchPath` become non-self references, and then the CA is wrong!
                // Instead we "corrupt" the path at `scratchPath` so when we later relocate it it's thereby uncorrupted, the opposite of the above problem.
```

Reply:
Indeed, `res.first` is a nicer option (I kept the `schatchPath != newInfo0.path` condition too because I think it's useful to prevent rewriting CA derivations).

Reply:
Looks good to me!